### ISSUE TITLE
imgui:x86-windows should be shared library

### ISSUE THREAD

**Issue by polelf** [microsoft/vcpkg#15501]
imgui:x86-windows should be shared library not static library

**Comment by JackBoosY** [comment]
Nope, according to the official documentation, imgui should be used as a header-only library, because all its symbols must be usable in custom projects (all symbols are not decleared as export in all platform).
At present, in vcpkg, imgui is only built as a static library for the same reason as above. The static library can export all symbols.

**Comment by polelf** [comment]
ogre component ogreBites ogreOverlay both use imgui 
static imgui library cause ogreBites.dll ogreOverlay.dll two ImGuiIO instances  to be generated

**Comment by polelf** [comment]
ogre component ogreBites ogreOverlay both use imgui
static imgui library cause ogreBites.dll ogreOverlay.dll two ImGuiIO instances to be generated but ogre engine just need one ImGuiIO

**Comment by JackBoosY** [comment]
cc @ras0219 @ras0219-msft

**Comment by Luke-Filewalker** [comment]
The reason why there are two instances is probably because of the situation JackBoosY explained. The imgui instances used in the dlls are not exported, therefore each ogre dll does not know of the other one and creates a new one. The only fix for this is to make them static or somehow export/pass the imgui data between them.

**Comment by polelf** [comment]
to build ogre samples Browser ImguiDemo, You will find all the problems

**Comment by PhoebeHui** [comment]
We suggest to build them statically, see https://github.com/microsoft/vcpkg/pull/15063#issuecomment-749991792 for the background.

**Comment by ras0219-msft** [comment]
imgui cannot be built as a shared library, so the only way to fix this issue is to force ogre to be static if it is built against imgui.

**Comment by sfhacker** [comment]
We are new to this project (imgui) so .... bear with us.
We have built imgui as a shared library (found an online contribution) and had no issues so far.
(Windows 10 x64)

![image](https://user-images.githubusercontent.com/25817529/106343026-830c5600-62f7-11eb-8f54-d233a64a70e6.png)


![image](https://user-images.githubusercontent.com/25817529/106342922-011c2d00-62f7-11eb-8109-10f1b1b249ab.png)

Just wondering how we can further test this project to make sure it works as expected?

Cheers.

**Comment by polelf** [comment]
The problem is when using ogre imguiOverlay

Use static ogre to avoid this problem

**Comment by BillyONeal** [comment]
> We have built imgui as a shared library (found an online contribution) and had no issues so far.

It's likely that this isn't something that is part of imgui itself rather than a DLL someone made attempting to export imgui's functionality. If it isn't something upstream supports/understands we generally don't want to make such changes. See maintainer policy "[Do not implement features in patches](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md#do-not-implement-features-in-patches)"

**Comment by ocornut** [comment]
Hello,

Author of dear imgui here.

Reasons why we discourage building as DLL:
- dear imgui does not provide any backward or forward binary compatibility by design. Using DLL puts in a path where it becomes much more likely that you use a version of Dear ImGui that you didn't build yourself and trip on ABI changes.
- ideal use involve compile-time settings such as defining implicit conversion from/to low-level math types to user's native maths types. While this is completely orthogonal to using DLL, there is a tendency where DLL are built far from end-user eyes, making it less discoverable and attractive to use a custom-build.
- dear imgui has a large API surface and advanced UI requires many function calls (10k function calls every 1/60 seconds is common, 100k is not unseen in edge cases) and we want to discourage people using anything that incur small overhead because at the end of day it'll be reflected in the app performances.

TL;DR; It can be built as a shared library we just generally discourage it and promote using static libraries and keeping the build process in sight and configurable.

**Comment by JackBoosY** [comment]
@ocornut  Thanks for reply, but I think at least imgui should support cmake.

**Comment by JackBoosY** [comment]
We hope your question was answered to your satisfaction; if it wasn't, you can reopen with more info.
